### ISSUE TITLE
Move HTTP response header validation to node

### ISSUE THREAD

**Issue by tepez** [hapijs/hapi#3065]
I recently got a `Header value cannot contain or convert into non-ascii characters: content-disposition` error after sending an attachment response for a file named `café.pdf`.

The error is thrown in [here](https://github.com/hapijs/hapi/blob/master/lib/response.js#L132) because the ASCII code of `é` is `233` and only codes smaller than `127` are allowed.

This is not coherent with the [content-disposition](https://github.com/jshttp/content-disposition) module which [allows](https://github.com/jshttp/content-disposition/blob/master/index.js#L81) up to code `255`.

I opened the issue here. If you feel hapi is doing the correct thing, I'll open it with content-disposition.

This check was added in this [commit](https://github.com/hapijs/hapi/commit/88539856f30a6ade6173b129af9dc57c72db4d88). @jefflembeck since it's your commit, could you please have a look?

**Comment by mtharrison** [comment]
255 would seem to be the correct value to support the entire ISO-8859-1 charset.

**Comment by tepez** [comment]
For reference: since 4.3.0 node's http module checks for invalid headers characters (itroduced in [this commit](https://github.com/nodejs/node/commit/cf2b714b02)) and it [allows](https://github.com/nodejs/node/blob/master/lib/_http_common.js#L246) characters in the range of: `9, [32-127], [128-255]`.

**Comment by jefflembeck** [comment]
This commit was originally introduced to prevent HTTP Response splitting in hapi apps on node 0.12 and 4 before 4.3.

This can likely be cleaned up by grabbing the same range that 4.3.0 allows

**Comment by davidkaminsky** [comment]
Is there a work-around at all for this? I'm running into it on hapi 13.4.0 and node 6.2.0.

**Comment by mtharrison** [comment]
The only "workaround" I can think of is to percent-encode (using `encodeURIComponent()`) any header fields containing chars outside of the printable ascii set. Not sure what is the right way to go here:

> Historically, HTTP has allowed field content with text in the
   ISO-8859-1 charset [ISO-8859-1], supporting other charsets only
   through use of [RFC2047] encoding.  In practice, most HTTP header
   field values use only a subset of the US-ASCII charset [USASCII].
   Newly defined header fields SHOULD limit their field values to
   US-ASCII octets.  A recipient SHOULD treat other octets in field
   content (obs-text) as opaque data.

Source: https://tools.ietf.org/html/rfc7230#section-3.2.4

**Comment by mtharrison** [comment]
@hueniverse would you support changing the header key/value character range to match what node does?

There are 2 issues when our validation differs to what happens in `res.setHeader()`:

1. We disallow valid characters (see original comment)
2. We allow some invalid characters (such as [ and ])

```js
server.route({
    method: 'GET',
    path: '/',
    handler: function (request, reply) {

        return reply().header('[header]', 'value');
    }
});
```

hapi won't throw on the above, but it will throw once it gets to node:

```
Debug: internal, implementation, error
    TypeError: Trailer name must be a valid HTTP Token ["[header]"]
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:347:13)
    ...
```

**Comment by kanongil** [comment]
If https://github.com/hapijs/hapi/pull/3178 is merged, this can be handled by simply removing the check from hapi, trusting node to throw on these headers. Given that all supported versions of node already make this check, this should be fine, right? Or do we support all v4.x implementations?

**Comment by hueniverse** [comment]
I Rather leave this up to node. If the security issue is not longer there, we can remove it from hapi.

**Comment by mtharrison** [comment]
The strict validation wasn't added to node until 4.3.0. @kanongil not sure about your question regarding which 4.x.x versions we support (I just assumed all). 

@jefflembeck are you able to confirm if the underlying security issue that prompted you to add the check has been fixed in node core >= 4.0.0?

**Comment by hueniverse** [comment]
We only support latest 4.x and 6.x. We can reflect that in the package.json file
